File: python/268-Missing Number.py
from typing import List

class Solution:
    def missingNumber(self, nums: List[int]) -> int:
        "Hashset, Time: O(n), Space: O(n)"
        n = len(nums)
        s = set(nums)
        for i in range(n+1):
            if i not in s:
                return i
        return 0
    # def missingNumber(self, nums: List[int]) -> int:
    #     "Math, Time: O(n), Space: O(1)"
    #     n = len(nums)
    #     expected_sum = n*(n+1) // 2
    #     actual_sum = sum(nums)
    #     return expected_sum - actual_sum

nums = [3,0,1]
result = Solution().missingNumber(nums)
assert result == 2

nums = [0,1]
result = Solution().missingNumber(nums)
assert result == 2

nums = [9,6,4,2,3,5,7,0,1]
result = Solution().missingNumber(nums)
assert result == 8

nums = [0]
result = Solution().missingNumber(nums)
assert result == 1

File: cpp/268. Missing Number.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
// Math, Time: O(n), Space: O(1)
public:
    int missingNumber(vector<int>& nums) {
        int n = nums.size();
        int expected = (0+n)*(n+1) / 2;
        int sum = accumulate(nums.begin(), nums.end(), 0);
        return expected - sum;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vi nums; int ans;

    solution = Solution();
    nums = {3,0,1};
    ans = solution.missingNumber(nums);
    assert(ans==2);

    solution = Solution();
    nums = {0,1};
    ans = solution.missingNumber(nums);
    assert(ans==2);

    solution = Solution();
    nums = {9,6,4,2,3,5,7,0,1};
    ans = solution.missingNumber(nums);
    assert(ans==8);
}